### ISSUE TITLE
Wrong usage of Cond.Signal()

### ISSUE THREAD

**Issue by kmfukuyama** [ochinchina/supervisord#223]
## Issue

- Cond.Signal() may be executed before Cond.Wait().

https://github.com/ochinchina/supervisord/blob/b857851be6fdf0f2038cd8127bbcec9e0c46bb01/supervisor.go#L537-L544

- Cond.Signal is used in 4 places.

## Solution

- Use channel or execute cond.L.Lock() before executing cond.Signal.

---

- Thank you for creating this repository!!

**Comment by kmfukuyama** [comment]
similar code

``` golang
package main

import (
	"fmt"
	"sync"
	"time"
)

func main() {

	cond := sync.NewCond(&sync.Mutex{})
	cnt := 0
	for {
		cnt++

		cond.L.Lock()

		go callback(func() {
			//cond.L.Lock()
			cond.Signal()
			//cond.L.Unlock()
		})

		cond.Wait()
		cond.L.Unlock()

		fmt.Printf("%d\n", cnt)

		time.Sleep(time.Millisecond * 10)
	}
}

func callback(f func()) {
	f()
}
```

Executing this code will cause a deadlock.
Delete A and it works fine.

**Comment by prashanthjbabu** [comment]
@kmfukuyama  I'm seeing an issue where sometimes when supervisord starts up , it doesnt start all the required processes. This happens randomly . Could this be due to the above issue?

**Comment by kmfukuyama** [comment]
@prashanthjbabu 
I don't think this fix will cause the issue you are facing.
However, if b857851be6fdf0f2038cd8127bbcec9e0c46bb01 doesn't cause the problem, and 2d043f1d9aeae61d17af79cba9fc08a7c8d1c166 commit only does, then it's possible that I made a mistake.

If the issue occurs regardless of the config values, I think the issue is very important and should be fixed quickly.
Would you please create a new github issue and provide details?

**Comment by prashanthjbabu** [comment]
@kmfukuyama Thanks for your reply ! I'm seeing the issue in Tag `v0.6.8` which doesn't have this commit . What I wanted to know is if your commit would have fixed this issue? Since according to your explanation , if `cond.Signal()` is hit before `cond.Wait()` , then we will be waiting at `cond.Wait()` forever and that might cause `startHTTServer()` to hang and not allow `startAutoStartPrograms()` to execute . So I feel your fix might solve the problem , but I just want to make sure!

**Comment by kmfukuyama** [comment]
@prashanthjbabu 
Very sorry. I misread the English text. I am writing with google translate...

> Since according to your explanation , if cond.Signal() is hit before cond.Wait() , then we will be waiting at cond.Wait() forever and that might cause startHTTServer() to hang and not allow startAutoStartPrograms() to execute .

Yes! That was fixed. 
The problem I faced was that supervisord freezes when I run "supervisord ctl reload".
Until now, running "supervisord ctl reload" dozens of times would freeze supervisord.
After this fix, supervisord did not freeze after running "supervisord ctl reload" 50,000 times.

The same code is executed at startup, so I think it has been fixed.
